network/server_discovery.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/0xPolygon/polygon-edge/network/proto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	kb "github.com/libp2p/go-libp2p-kbucket"
+	rawGrpc "google.golang.org/grpc"
+	"math/big"
+	"time"
+)
+
+var (
+	errPeerDisconnected = errors.New("peer disconnected before the discovery client was initialized")
+)
+
+// GetRandomBootnode fetches a random bootnode that's currently
+// NOT connected, if any
+func (s *Server) GetRandomBootnode() *peer.AddrInfo {
+	nonConnectedNodes := make([]*peer.AddrInfo, 0)
+
+	for _, v := range s.bootnodes.getBootnodes() {
+		if !s.hasPeer(v.ID) {
+			nonConnectedNodes = append(nonConnectedNodes, v)
+		}
+	}
+
+	if len(nonConnectedNodes) > 0 {
+		randNum, _ := rand.Int(rand.Reader, big.NewInt(int64(len(nonConnectedNodes))))
+
+		return nonConnectedNodes[randNum.Int64()]
+	}
+
+	return nil
+}
+
+// GetBootnodeConnCount fetches the number of active bootnode connections [Thread safe]
+func (s *Server) GetBootnodeConnCount() int64 {
+	return s.bootnodes.getBootnodeConnCount()
+}
+
+// getProtoStream returns an active protocol stream if present, otherwise
+// it returns nil
+func (s *Server) getProtoStream(protocol string, peerID peer.ID) *rawGrpc.ClientConn {
+	s.peersLock.Lock()
+	defer s.peersLock.Unlock()
+
+	connectionInfo, ok := s.peers[peerID]
+	if !ok {
+		return nil
+	}
+
+	return connectionInfo.getProtocolStream(protocol)
+}
+
+// NewDiscoveryClient returns a new or existing discovery service client connection
+func (s *Server) NewDiscoveryClient(peerID peer.ID) (proto.DiscoveryClient, error) {
+	// Temporary dials are never added to the peer store,
+	// so they have a special status when doing discovery
+	isTemporaryDial := s.IsTemporaryDial(peerID)
+
+	// Check if there is a peer connection at this point in time,
+	// as there might have been a disconnection previously
+	if !s.isConnected(peerID) && !isTemporaryDial {
+		return nil, errPeerDisconnected
+	}
+
+	// Check if there is an active stream connection already
+	if protoStream := s.getProtoStream(common.DiscProto, peerID); protoStream != nil {
+		return proto.NewDiscoveryClient(protoStream), nil
+	}
+
+	// Create a new stream connection and return it
+	protoStream, err := s.newProtoConnection(common.DiscProto, peerID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Discovery protocol streams should be saved,
+	// since they are referenced later on,
+	// if they are not temporary
+	if !isTemporaryDial {
+		s.saveProtocolStream(common.DiscProto, protoStream, peerID)
+	}
+
+	return proto.NewDiscoveryClient(protoStream), nil
+}
+
+// saveProtocolStream saves the protocol stream to the peer
+// protocol stream reference [Thread safe]
+func (s *Server) saveProtocolStream(
+	protocol string,
+	stream *rawGrpc.ClientConn,
+	peerID peer.ID,
+) {
+	s.peersLock.Lock()
+	defer s.peersLock.Unlock()
+
+	connectionInfo := s.peers[peerID]
+	connectionInfo.addProtocolStream(protocol, stream)
```

Comment:
There might be nil panic

Reply:
We don't call it from an unsafe place anyways, but I've added a sanity check for the future:

[99f82f7](https://github.com/0xPolygon/polygon-edge/pull/343/commits/99f82f7bd75bc808d1a37c0dc90c37e19c904283)